### PR TITLE
Utils tarball extension

### PR DESCRIPTION
**Pull request by fogre**
Tarball creator extension for tmc-langs-util. Creates a tarball that can be sent to TMC-sandbox for testing.

The created tarball has been tested locally and should be similar to the one created by TMC-Server.

### CODE REVIEW

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
+package fi.helsinki.cs.tmc.langs.util.tarservice;
```
Comment (Redande) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java]: We might later include language-specific functionalities here, so maybe this file belongs to tmc-langs-framework

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+     * @param project        Project directory file
+     * @param targetLocation Location where the tar archive should be extracted to
+     * @throws IOException      Error!
+     * @throws ArchiveException Error!
+     */
+
+    private void createTarBall(File project, Path targetLocation)
+            throws IOException, ArchiveException {
+        FileOutputStream tarOut;
+        if (targetLocation.getFileName().toString().contains(".tar")) {
```
Comment (Redande) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java]:
```java
if (targetLocation.endsWith(".tar"))
```

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            tarOut = new FileOutputStream(targetLocation.toString());
+        } else {
+            tarOut = new FileOutputStream(targetLocation.toString() + ".tar");
```
Comment (Redande) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java]: Will anyone ever call this function with a `targetLocation` that doesn't end with `.tar`?

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+     */
+
+    private void addFilesToTarBall(File folder, TarArchiveOutputStream tar,
+                                   int lengthOfPath) throws FileNotFoundException, IOException {
+        for (final File file : folder.listFiles()) {
+            if (file.isDirectory()) {
+                addFilesToTarBall(file, tar, lengthOfPath);
+            } else {
+                TarArchiveEntry entry;
+                entry = new TarArchiveEntry(file.toString().substring(lengthOfPath));
```
Comment (Redande) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java]: Once you have done the `newDirectoryStream` refactoring, the `file` should be an instance of `Path`. Then you can accomplish relativizing simply by:
```java
basePath.relativize(file);
```
Note that you also need to pass `basePath` from the caller instead of the base path's length.

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                entry.setSize(file.length());
+                tar.putArchiveEntry(entry);
+                FileInputStream fis = new FileInputStream(file);
+                IOUtils.copy(fis, tar);
+                fis.close();
+                tar.closeArchiveEntry();
+            }
+
```
Comment (Redande) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java]: Unnecessary extra line

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        }
+    }
+
```
Comment (Redande) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java]: Unnecessary extra line

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            tarOut = new FileOutputStream(targetLocation.toString());
+        } else {
+            tarOut = new FileOutputStream(targetLocation.toString() + ".tar");
+        }
+        TarArchiveOutputStream aos = new TarArchiveOutputStream(tarOut);
+        aos.setLongFileMode(LONGFILE_POSIX);
+        addFilesToTarBall(project, aos, project.getAbsolutePath().length());
+        aos.finish();
+        aos.close();
+        tarOut.close();
```
Comment (ljleppan) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java]: Use [Try with Resource](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) to ensure these close properly even if an exception is raised.

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                addFilesToTarBall(file, tar, lengthOfPath);
+            } else {
+                TarArchiveEntry entry;
+                entry = new TarArchiveEntry(file.toString().substring(lengthOfPath));
+                entry.setSize(file.length());
+                tar.putArchiveEntry(entry);
+                FileInputStream fis = new FileInputStream(file);
+                IOUtils.copy(fis, tar);
+                fis.close();
+                tar.closeArchiveEntry();
```
Comment (ljleppan) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java]: Ensure these close properly (try with resources?)

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java (hunk 7 above)
Comment (fogre) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java]: I need to change the project source/binary format to include try with resource. If that is not ok, I can use finally and check that they are properly closed.

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java (hunk 1 above)
Comment (fogre) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java]: We can move it there if necessary.

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+     *
+     * @param project        Project directory file
+     * @param targetLocation Location where the tar archive should be extracted to
+     * @throws IOException      Error!
+     * @throws ArchiveException Error!
+     */
+    private void createTarBall(Path project, Path targetLocation)
+            throws IOException, ArchiveException {
+        try (FileOutputStream tarOut = new FileOutputStream(targetLocation.toString())) {
+            try (TarArchiveOutputStream aos = new TarArchiveOutputStream(tarOut)) {
```
Comment (ljleppan) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java]: **Comment, not a request for changes:** You could combine these two `try`lines into a single line as 
```
try (TarArchiveOutputStream aos = new TarArchiveOutputStream(new FileOutputStream(targetLocation.toString()))) {
```

The `try`-clause then calls `close()` for `aos`, which then calls `close()` of the `FileOutputStream` explicitly. But that's getting a bit wordy and I'm not sure I like it.

Alternatively you could have multiple statements in a single try-with-resources statement, like this:
```
try (FileOutputStream tarOut = new FileOutputStream(targetLocation.toString());
      TarArchiveOutputStream aos = new TarArchiveOutputStream(tarOut)) {
```

But that's also kinda ugly and also ends up calling `close`on `tarOut` twice (once implicitly thru `aos`, once explicitly in the (hidden) `finally`block).

**The current implementation is fine in my books**, as I find all three possible solutions sort-of-verbose.  

If you are interested in these kinds of things, there's some discussion on a similar case here: https://stackoverflow.com/questions/12552863/correct-idiom-for-managing-multiple-chained-resources-in-try-with-resources-bloc

File: tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    private void addFilesToTarBall(Path folder, TarArchiveOutputStream tar,
+                                   Path basePath) throws FileNotFoundException, IOException {
+        for (Path path : Files.newDirectoryStream(folder)) {
+            if (Files.isDirectory(path)) {
+                addFilesToTarBall(path, tar, basePath);
+            } else {
+                TarArchiveEntry entry = new TarArchiveEntry(basePath.relativize(path).toString());
+                entry.setSize(path.toFile().length());
+                tar.putArchiveEntry(entry);
+                try (FileInputStream fis = new FileInputStream(path.toFile())) {
```
Comment (ljleppan) [inline tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java]: **Comment, not request for changes**:

`Files.newInputStream(Path path)` would be a valid way of doing this without the `Path` -> `File` transformation.
